torch_xla/debug/metrics_compare_utils.py

```diff
@@ -0,0 +1,184 @@
+    {
+      'CompileTime__Accumulator_sec': [50, None, 80],  # missing from report2
+      'InboundData__TotalSamples': [11, 12, 11],
+      'CreateCompileHandles__Value': [20, 20, 19],
+      ...
+    }
+  """
+  data_points = collections.defaultdict(lambda: [None] * len(metrics_reports))
+  for report_index in range(len(metrics_reports)):
+    parsed_report = _parse_metrics_report(metrics_reports[report_index])
+    for metric_name in parsed_report:
+      data_points[metric_name][report_index] = parsed_report.get(
+          metric_name, None)
+  return data_points
+
+
+def _compute_aggregates(data_points):
+  # Compute mean and standard deviation for each key in data_points.
+  aggregates = {}
+  for k, v_list in data_points.items():
+    v_list = [v for v in v_list if v is not None]
+    if v_list:
+      aggregates[k] = {'mean': mean(v_list), 'stddev': std(v_list)}
+  return aggregates
+
+
+def compare_metrics(data_points, metrics_report,
+    config={'base_expression': 'v <= v_mean + (v_stddev * 2.0)'}):
+  """Compare metrics_report to historical averages and report differences.
+
+  Args:
+    data_points(dict): Output of get_data_points_from_metrics_reports().
+    metrics_report(string): Output of torch_xla metrics_report(). These will
+      be compared with the aggregates of raw_data_points.
+    config(dict): Configuration for metrics comparison. You can supply a
+      per-metric expression for any aggregated metric name, otherwise
+      base_expression will be used. Example config:
+      {
+        'base_expression': 'v <= v_mean + (v_stddev * 2.0)',
+        'HigherIsBetter_Accumulator_sec_expression':
+            'v >= v_mean - (v_stddev * 3.0)',
+        ...
+      }
+
+      Expressions should be Python expressions that return True or False
+      depending on the values of:
+        v: the value of this metric based on `metrics_report`.
+        v_mean: the mean value of this metric based on `data_points`.
+        v_stddev: the stddev of this metric based on `data_points`.
+
+  Returns:
+    Metrics difference report (string). For any metric that differed between
+    metrics_report and the aggregates of raw_data_points, this report will have
+    1 line reporting the difference.
+  """
+  parsed_report = _parse_metrics_report(metrics_report)
+  means_and_stddevs = _compute_aggregates(data_points)
+
+  difference_report = ''
+  for k, v in sorted(parsed_report.items()):
+    if k not in means_and_stddevs:
+      # Alert if we have new ops falling back to CPU.
+      if k.startswith('aten::'):
+        difference_report += 'Found new aten counter: {}: {}\n'.format(k, v)
+    else:
+      v_mean = means_and_stddevs[k]['mean']
+      v_stddev = means_and_stddevs[k]['stddev']
+
+      # Expression can be overridden for individual metrics.
+      expression = config.get('{}_expression'.format(k))
+      if expression is None:
+        expression = config['base_expression']
+
+      if eval(expression) == False:
```

Comment:
For security reasons it is better to build a dedicated, limited, dictionary to pass to `eval()`.
Is is also cleaner instead of `eval()` magically picking up from local/global frames.

Reply:
Ah yes, great idea. I wasn't familiar with the local/global args for eval()